falcon/cyutil/uri.pyx

```diff
@@ -242,12 +242,12 @@ cdef cy_parse_query_string(unsigned char* data, Py_ssize_t length,
 
 def parse_query_string(unicode query_string not None, bint keep_blank=False,
                        bint csv=True):
-    cdef bytes byte_string = query_string.encode('utf-8')
```

Comment:
It might be that this change is actually a de-optimization, or at least a no-op.
Let me check.

Reply:
Ah, good point. Now that you mention it I vaguely remember talking about this a while back. Better double-check.

Reply:
Dumping here for reference

```c
  /* "testutf8.pyx":2
 * def do_something_with_string1(unicode input_string not None):
 *     cdef bytes distinguishable_name = input_string.encode('utf-8')             # <<<<<<<<<<<<<<
 *     return len(distinguishable_name)
 * 
 */
  __pyx_t_1 = PyUnicode_AsUTF8String(__pyx_v_input_string); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 2, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_distinguishable_name = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;
```

```c
  /* "testutf8.pyx":7
 * 
 * def do_something_with_string2(unicode input_string not None):
 *     cdef bytes distinguishable_name = input_string.encode()             # <<<<<<<<<<<<<<
 *     return len(distinguishable_name)
 */
  __pyx_t_1 = PyUnicode_AsEncodedString(__pyx_v_input_string, NULL, NULL); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 7, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_1);
  __pyx_v_distinguishable_name = ((PyObject*)__pyx_t_1);
  __pyx_t_1 = 0;
```

Reply:
@kgriffs I'm getting a very very tiny, but nevertheless reliably reproducible, performance regression with your version.
```python
>>> timeit.timeit(lambda: do_something_with_string1(s), number=int(1e8))
7.346965251000029
>>> timeit.timeit(lambda: do_something_with_string2(s), number=int(1e8))
7.4253033239999695
>>> timeit.timeit(lambda: do_something_with_string1(s), number=int(1e8))
7.352099975000101
>>> timeit.timeit(lambda: do_something_with_string2(s), number=int(1e8))
7.417692562999946
>>> timeit.timeit(lambda: do_something_with_string1(s), number=int(1e8))
7.380943102999936
>>> timeit.timeit(lambda: do_something_with_string2(s), number=int(1e8))
7.449390696999899
>>> s
'This is an ASCII string.'
```

In that light, I would suggest reverting `uri.pyx`.

Reply:
my guess is that `PyUnicode_AsEncodedString` checks what is the default encoding of the system and then calls `PyUnicode_AsUTF8String`.